File: core/src/services/osrm-client.ts
import type { LatLng } from "./haversine";

export interface OsrmRoute {
  distanceMeters: number;
  durationSeconds: number;
}

export type OsrmFailureReason =
  | "unreachable"
  | "timeout"
  | "no_route"
  | "error_status"
  | "malformed_response";

export type OsrmResult =
  | { ok: true; route: OsrmRoute }
  | { ok: false; reason: OsrmFailureReason; detail: string };

export interface OsrmClientConfig {
  osrmUrl: string;
  timeoutMs: number;
}

/**
 * Calls a real OSRM instance's GET /route/v1/driving (docs/osrm-routing.md) with a hard timeout,
 * returning a discriminated result rather than throwing — same reusable fallback pattern as
 * ml-eta-client.ts's fetchMlEta (Phase 10).
 *
 * One important, verified difference from ml-service's convention: OSRM signals a routing
 * failure (e.g. a point that can't be snapped to any road within its search radius) with HTTP
 * 400 and a JSON body like `{"code":"NoSegment","message":"..."}` — NOT a 200 with an in-body
 * error code. So the body is always parsed as JSON first, regardless of HTTP status, and a
 * recognizable non-"Ok" `code` field is treated as "no_route" before falling back to
 * status-code-based classification for genuinely malformed/unexpected responses.
 */
export async function fetchOsrmRoute(
  from: LatLng,
  to: LatLng,
  cfg: OsrmClientConfig,
): Promise<OsrmResult> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), cfg.timeoutMs);

  const url =
    `${cfg.osrmUrl}/route/v1/driving/${from.lng},${from.lat};${to.lng},${to.lat}` +
    `?overview=false`;

  let response: Response;
  try {
    response = await fetch(url, { signal: controller.signal });
  } catch (err) {
    if (err instanceof Error && err.name === "AbortError") {
      return { ok: false, reason: "timeout", detail: `no response within ${cfg.timeoutMs}ms` };
    }
    const message = err instanceof Error ? err.message : String(err);
    return { ok: false, reason: "unreachable", detail: message };
  } finally {
    clearTimeout(timer);
  }

  let body: unknown;
  try {
    body = await response.json();
  } catch {
    return response.ok
      ? { ok: false, reason: "malformed_response", detail: "response body was not valid JSON" }
      : { ok: false, reason: "error_status", detail: `osrm responded ${response.status}` };
  }

  const code = isRecord(body) ? body.code : undefined;
  if (typeof code === "string" && code !== "Ok") {
    const message = isRecord(body) ? body.message : undefined;
    return {
      ok: false,
      reason: "no_route",
      detail: typeof message === "string" ? message : `osrm code=${code}`,
    };
  }

  if (!response.ok) {
    return { ok: false, reason: "error_status", detail: `osrm responded ${response.status}` };
  }

  const route = parseRoute(body);
  if (!route) {
    return {
      ok: false,
      reason: "malformed_response",
      detail: `unexpected response shape: ${JSON.stringify(body)}`,
    };
  }

  return { ok: true, route };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function parseRoute(body: unknown): OsrmRoute | null {
  if (!isRecord(body) || !Array.isArray(body.routes) || body.routes.length === 0) return null;
  const first: unknown = body.routes[0];
  if (!isRecord(first)) return null;

  const distanceMeters = first.distance;
  const durationSeconds = first.duration;
  if (typeof distanceMeters !== "number" || !Number.isFinite(distanceMeters) || distanceMeters < 0) {
    return null;
  }
  if (
    typeof durationSeconds !== "number" ||
    !Number.isFinite(durationSeconds) ||
    durationSeconds < 0
  ) {
    return null;
  }

  return { distanceMeters, durationSeconds };
}

File: ml-service/app/ml/osrm_client.py
"""Real road-network routing via OSRM (Phase 15, docs/osrm-routing.md) — used by
scripts/train_model.py to compute a per-trip `osrm_duration_seconds`/`osrm_distance_meters`
feature for the retrained model. Mirrors core/src/services/osrm-client.ts's design (a
discriminated result instead of raising, so callers can distinguish failure modes) and,
critically, its corrected parsing order: OSRM signals a routing failure (e.g. a point that can't
be snapped to a road, `NoSegment`/`NoRoute`) via **HTTP 400** with a JSON body
`{"code": "NoSegment", "message": "..."}` — not a 200-with-error-code the way this same service's
own /predict-eta does. So the body is always parsed as JSON first, regardless of HTTP status, and
a recognizable non-"Ok" `code` is treated as "no_route" before falling back to
status/shape-based classification. Verified against a real running OSRM container, not assumed —
see docs/osrm-routing.md.
"""

from dataclasses import dataclass
from typing import Any

import httpx


@dataclass
class OsrmRoute:
    distance_meters: float
    duration_seconds: float


@dataclass
class OsrmResult:
    ok: bool
    route: OsrmRoute | None = None
    # "unreachable" | "timeout" | "no_route" | "error_status" | "malformed_response"
    reason: str | None = None
    detail: str | None = None


def fetch_osrm_route(
    osrm_url: str,
    pickup_lat: float,
    pickup_lng: float,
    dropoff_lat: float,
    dropoff_lng: float,
    timeout_seconds: float,
) -> OsrmResult:
    url = (
        f"{osrm_url}/route/v1/driving/"
        f"{pickup_lng},{pickup_lat};{dropoff_lng},{dropoff_lat}"
        f"?overview=false"
    )

    try:
        response = httpx.get(url, timeout=timeout_seconds)
    except httpx.TimeoutException as err:
        return OsrmResult(ok=False, reason="timeout", detail=str(err) or "timed out")
    except httpx.RequestError as err:
        return OsrmResult(ok=False, reason="unreachable", detail=str(err) or "request failed")

    try:
        body = response.json()
    except ValueError:
        if response.is_success:
            return OsrmResult(
                ok=False,
                reason="malformed_response",
                detail="response body was not valid JSON",
            )
        return OsrmResult(
            ok=False, reason="error_status", detail=f"osrm responded {response.status_code}"
        )

    code = body.get("code") if isinstance(body, dict) else None
    if isinstance(code, str) and code != "Ok":
        message = body.get("message") if isinstance(body, dict) else None
        detail = message if isinstance(message, str) else f"osrm code={code}"
        return OsrmResult(ok=False, reason="no_route", detail=detail)

    if not response.is_success:
        return OsrmResult(
            ok=False, reason="error_status", detail=f"osrm responded {response.status_code}"
        )

    route = _parse_route(body)
    if route is None:
        return OsrmResult(
            ok=False,
            reason="malformed_response",
            detail=f"unexpected response shape: {body}",
        )

    return OsrmResult(ok=True, route=route)


def _parse_route(body: Any) -> OsrmRoute | None:
    if not isinstance(body, dict):
        return None
    routes = body.get("routes")
    if not isinstance(routes, list) or len(routes) == 0:
        return None
    first = routes[0]
    if not isinstance(first, dict):
        return None

    distance = first.get("distance")
    duration = first.get("duration")
    if not isinstance(distance, int | float) or isinstance(distance, bool) or distance < 0:
        return None
    if not isinstance(duration, int | float) or isinstance(duration, bool) or duration < 0:
        return None

    return OsrmRoute(distance_meters=float(distance), duration_seconds=float(duration))
